src/js/angular/guides/tour-lib-services/shepherd.service.js

```diff
@@ -0,0 +1,543 @@
+        buttons.push(this._getPreviousButton(guide, previousStepDescription, currentStepDescription));
+        buttons.push(this._getNextButton(guide, currentStepDescription, nexStepDescription));
+        buttons.push(this._getPauseButton(guide));
+        step.buttons = buttons;
+        return step;
+    }
+
+    this._getPauseButton = (guide) => {
+        return this._getButton($translate.instant('guide.button.pause'), () => this._pauseGuide(guide), true);
+    }
+
+    /**
+     * Pauses current ran guide.
+     */
+    this._pauseGuide = (guide) => {
+        guide.hide();
+        this._getStepWhichCanBePaused(Shepherd.activeTour);
+        LocalStorageAdapter.set(GUIDE_PAUSE, true);
+        if (this.onPause()) {
+            this.onPause();
+        }
+    }
+
+    this._getStepWhichCanBePaused = (tour) => {
+        const currentStep = tour.getCurrentStep();
+        if (currentStep) {
+            if (currentStep.options.canBePaused) {
+                return currentStep;
+            }
+            tour.back();
+            tour.hide();
+            const previousStep = tour.getCurrentStep();
+            // checks if ids are equal this mean first step is reached.
+            if (currentStep.id === previousStep.id) {
+                return currentStep;
+            } else {
+                this._getStepWhichCanBePaused(tour);
+            }
+        }
+    }
+
+    /**
+     * Creates a previous button.
+     * @param guide - the guide.
+     * @param previousStepDescription - previous step description.
+     * @param currentStepDescription - processed step description.
+     * @returns created step.
+     * @private
+     */
+    this._getPreviousButton = (guide, previousStepDescription, currentStepDescription) => {
+        const text = $translate.instant('guide.button.previous');
+        const action = this._getPreviousButtonAction(guide, previousStepDescription, currentStepDescription);
+        const button = this._getButton(text, action);
+        button.disabled = !previousStepDescription;
+        return button;
+    }
+
+    this._getPreviousButtonAction = (guide, previousStepDescription, currentStepDescription) => {
+        if (!!previousStepDescription) {
+            return () => {
+                if ('clickable' === previousStepDescription.type && angular.isFunction(previousStepDescription.onPreviousClick)) {
+                    previousStepDescription.onPreviousClick();
+                } else if (previousStepDescription.forceReload || previousStepDescription.url && '' !== previousStepDescription.url && previousStepDescription.url !== currentStepDescription.url) {
+                    $location.path(previousStepDescription.url);
+                }
+                GuideUtils.waitFor(previousStepDescription.elementSelector, previousStepDescription.maxWaitTime)
+                    .then(() => guide.back())
+                    .catch(() => guide.cancel());
+            }
+        }
+        return () => {};
```

Comment:
Shouldn't we notify for such cases where there is no step description? What is expected for such button to do?

Reply:
This was just empty function when there is not nextStepDescription and next button is disabled. I removed it because now next button not added in this case. Same for previous button.